resources/type-system/inference.md

```diff
@@ -970,6 +970,391 @@ with respect to `L` under constraints `C0`
+the presence of coercions, this can sometimes be confusing, since a given
+expression might be associated with more than one expression artifact. For
+example in the code `dynamic d = ...; int i = d;`, the expression `d` is
+associated with two expression artifacts: one which is the result of reading the
+value of the variable `d`, and one which is the result of casting that value to
+the type `int`. The former artifact has static type `dynamic`; the latter has
+static type `int`. Since this document details precisely when and how coercions
+occur, and makes arguments about soundness of expressions that might involve
+coercions, it is useful to be precise by associating static types with the
+expression artifacts rather than the expressions themselves._
+
+## Soundness guarantees
+
+Along with the notion of _expression soundness_ described above, there are
+several other soundness guarantees made by Dart:
+
+- It is guaranteed that if a method, function, or constructor having return type
+  `T` is invoked, and the method completes with a value `v`, then `v` will be an
+  instance of `T` (with appropriate type substitutions, in the case of a generic
+  method). This guarantee is known as _return value soundness_.
+
+- It is guaranteed that if a method having static type `T` is torn off, then the
+  resulting value will be an instance of `T`. This guarantee is known as
+  _tear-off soundness_.
+
+- It is guaranteed that if a future is an instance of the type `Future<T>`, and
+  it completes with a value `v`, then `v` will be an instance of `T`. This
+  guarantee is known as _future soundness_.
+
+The rules below include informal sketches of a proof that each of the above
+soundness guarantees holds. These are non-normative, so they are typeset in
+_italics_.
+
+# Coercions
+
+_Coercion_ is a type inference step that transforms an expression artifact `m'`,
+and a desired static type `T`, into an expression artifact `m` whose static type
+is `T`.
+
+_Coercions are used in most situations where the existing spec calls for an
+assignability check._
+
+Coercion of an expression artifact `m'` to type `T` produces an expression
+artifact `m` with static type `T`, where `m` is determined as follows:
+
+- Let `T'` be the static type of `m'`.
+
+- Define `m` as follows:
+
+  - If `T' <: T`, then let `m` be an expression artifact whose runtime behavior
+    is as follows:
+
+    - Execute expression artifact `m'`, and let `o` be the resulting value. _By
+      execution soundness, `o` will be an instance of the type `T'`._
+
+    - `m` completes with the value `o`. _Expression soundenss follows from the
+      fact that since `T' <: T`, `o` must also be an instance of the type `T`._
+
+  - Otherwise, if `T'` is `dynamic`, then let `m` be an expression artifact
+    whose runtime behavior is as follows:
+
+    - Execute expression artifact `m'`, and let `o` be the resulting value. _By
+      execution soundness, `o` will be an instance of the type `T'`._
```

Comment:
I'm not sure if the non-normative text is saying anything meaningful here since `T'` is `dynamic`. Every value is an instance of `dynamic`. Maybe just remove it?

Reply:
Good point, done. Thanks for the suggestion.